addons/l10n_pk_edi/models/account_move.py

```diff
@@ -0,0 +1,568 @@
+            "hsCode": self._l10n_pk_edi_extract_digits(product_id.hs_code),
+            "productCode": product_id.code,
+            "productDescription": prd_desc.replace("\n", ""),
+            "rate": self._l10n_pk_round_value(line.price_unit or 0.0),
+            "uoM": line.product_uom_id.l10n_pk_code,
+            "quantity": self._l10n_pk_round_value(quantity),
+            "valueSalesExcludingST": self._l10n_pk_round_value(sale_value + line_discount),
+            "salesTaxApplicable": self._l10n_pk_round_value(sales_tax_total or 0.0),
+            "retailPrice": self._l10n_pk_round_value(line.price_unit or 0.0),
+            "salesTaxWithheldAtSource": 0.0,
+            "extraTax": self._l10n_pk_round_value(non_sales_tax_total or 0.0),
+            "furtherTax": 0.0,
+            "sroScheduleNo": None,
+            "fedPayable": 0.0,
+            "cvt": 0.0,
+            "withholdingIncomeTaxApplicable": self._l10n_pk_round_value(tax_withholdings.get('amount', 0.0)),
+            "whiT_1": 0.0,
+            "whiT_2": None,
+            "whiT_Section_1": None,
+            "whiT_Section_2": "0.0",
+            "totalValues": self._l10n_pk_round_value(sale_value + sales_tax_total),
+            "discount": self._l10n_pk_round_value(line_discount),
+            "sroItemSerialNo": (
+                product_id.l10n_pk_edi_schedule_code_id.schedule_code[-3:]
+                if product_id.l10n_pk_edi_schedule_code_id else False
+            ),
+            "stWhAsWhAgent": 0,
+        }
+        line_json[
+            "saleType" if move_id.is_sale_document() else "purchaseType"
+        ] = product_id.l10n_pk_edi_sale_type_id.sale_type
+        if reference_number := self._get_l10n_pk_edi_fbr_ref_number():
+            line_json["invoiceRefNo"] = reference_number
+        return line_json
+
+    def _l10n_pk_edi_generate_invoice_json(self):
+        def _get_buyer_seller():
+            if self.is_sale_document():
+                return self.partner_id, self.company_id.partner_id
+            else:  # purchase_document
+                return self.company_id, self.partner_id
+        tax_details = self._prepare_invoice_aggregated_taxes()
+        tax_details_per_record = tax_details.get("tax_details_per_record")
+        lines = self.invoice_line_ids.filtered(lambda line: line.display_type not in (
+            'line_note', 'line_section', 'rounding'
+        ))
+        global_discount_line = lines.filtered(self._l10n_pk_is_global_discount)
+        lines -= global_discount_line
+        # Calculate the total discount as the sum of the global discount and the line discounts
+        total_discount = (
+            sum(line.balance for line in global_discount_line) * (1 if self.is_inbound() else -1)
+        ) + sum((line.price_unit * line.quantity) * (line.discount / 100) for line in lines)
+        # Total Sale Value (Excluding Discount and Tax Amount)
+        total_sale_value = sum(line.price_unit * line.quantity for line in lines)
```

Comment:
```suggestion
        total_sale_value = sum(line.price_subtotal for line in lines)
```
@masm-odoo Apply it everywhere applicable

Thanks @jco-odoo

Reply:
@hamo-odoo @jco-odoo Thanks for the suggestion 😊. However, the `price_subtotal` field includes line discounts, which I want to exclude from my calculations. I need a way to accurately compute the subtotal without the discount and tax applied 🤔.